cdap-proto/src/main/java/co/cask/cdap/proto/RunRecord.java

```diff
@@ -123,6 +143,101 @@ public String toString() {
       ", stopTs=" + stopTs +
       ", status=" + status +
       ", properties=" + properties +
+      ", cluster=" + cluster +
       '}';
   }
+
+  /**
+   * @return Builder to create a RunRecord
+   */
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  /**
+   * @param runRecord existing record to copy fields from
+   * @return Builder to create a RunRecord, initialized with values from the specified existing record
+   */
+  public static Builder builder(RunRecord runRecord) {
+    return new Builder(runRecord);
+  }
+
+  /**
+   * Builder to create RunRecords.
+   *
+   * @param <T> type of builder
+   */
+  @SuppressWarnings("unchecked")
+  public static class Builder<T extends Builder> {
+    protected ProgramRunStatus status;
+    protected String pid;
+    protected Long startTs;
+    protected Long runTs;
+    protected Long stopTs;
+    protected Map<String, String> properties;
+    protected ProgramRunCluster cluster;
+
+    protected Builder() {
+      properties = new HashMap<>();
+    }
+
+    protected Builder(RunRecord other) {
+      status = other.getStatus();
+      pid = other.getPid();
+      startTs = other.getStartTs();
+      runTs = other.getRunTs();
+      stopTs = other.getStopTs();
+      properties = new HashMap<>(other.getProperties());
+      cluster = other.getCluster();
+    }
+
+    public T setStatus(ProgramRunStatus status) {
+      this.status = status;
+      return (T) this;
+    }
+
+    public T setRunId(String runId) {
+      this.pid = runId;
+      return (T) this;
+    }
+
+    public T setStartTime(long startTs) {
+      this.startTs = startTs;
+      return (T) this;
+    }
+
+    public T setRunTime(Long runTs) {
+      this.runTs = runTs;
+      return (T) this;
+    }
+
+    public T setStopTime(Long stopTs) {
+      this.stopTs = stopTs;
+      return (T) this;
+    }
+
+    public T setProperties(Map<String, String> properties) {
+      this.properties.clear();
+      this.properties.putAll(properties);
+      return (T) this;
+    }
+
+    public T setCluster(ProgramRunCluster cluster) {
+      this.cluster = cluster;
+      return (T) this;
+    }
+
+    public RunRecord build() {
+      if (pid == null) {
+        throw new IllegalArgumentException("Run record run id must be specified.");
+      }
+      if (startTs == null) {
+        throw new IllegalArgumentException("Run record start time must be specified.");
+      }
+      if (cluster == null) {
```

Comment:
should we also check if `programStatus` is null?

Reply:
yep, you're right. will do